### PR TITLE
Multiple sharelinks

### LINKED ISSUE #866: Multiple Share Links
Multiple share links is a new feature in oC10, that allows to create more than one public link over the same file/folder.

Current behaviour:

- The iOS mobile app only supports one public link. 
- The behaviour right now is showing the last one of the public links created on the server side. 
- If the link is edited in the iOS app, the name is removed from the link in server side  -> https://github.com/owncloud/core/issues/27622
- If the link is removed from the device, the previous one is displayed

It will require some actions:

- [x] UI changes to adapt to the new feature. As proposal, they can be handled in the same way as the internal sharing: "Add public links" button (or "+" symbol regarding https://github.com/owncloud/ios/pull/859) with an option to edit the public link options (name, allow editing if folder, expiration, password)
- [x] Update wordings "Public link" -> "Public Links", "Share with users.." -> "Users an groups"
- [x] Update  shareObj and parser in library
- [x] Update DB with new name parameter
- [x] Update capabilities with the new parameter "files_sharing" > "multiple"  in library (see https://github.com/owncloud/core/issues/27622 )
- [x] Update db for capability
- [X] Replace "i" icon in sharees cell with a disclosure - [I rather to keep it with the information icon that for iOS means new view where the user can edit, the disclosure icon it is used to show a navigation]
- [x] New public link cell will show the name of the link if exist and a disclosure to navigate to details
- [x] New button "public link icon" in public link cell that will allow share the link 
- [x] Move current settings options for public links to new view as share users edit options
- [x] Add a new field related that allow set up the link name. (set file name as a default)
- [x] Warning about public links, new parameter k_warning_sharing_public_link https://github.com/owncloud/enterprise/issues/1914
- [x] Get correct share link and show share activity view
- [X] Create new methods in library to support update new name parameter (deprecate old one) 
- [X] Create new methods in library to support create new links with all options (deprecate old one)
- [x] Add new library tests for the new methods
- [X] Handle the different cases when updating the public share password, allow uploads, name, expiration
- [x] Grant that password is required when server enforces it
- [x] Grant that expiration time is required when server enforces it
- [x] Ensure that expiration date is shown by default when server set up
- [x] Handle errors
- [X] Grant that share link not shown if cap is disabled
- [X] Enable or disable multiple links support depending on the corresponding capability (see https://github.com/owncloud/core/issues/27622 )
- [X] Hide new parameter name in create and update view for server minor 10 or not multiple cap enabled
- [X] Always show Warning about public share link with message "Anyone with the link has access to the file/folder"  with gray color https://github.com/owncloud/core/pull/27844/files
- [x] Shared public links view. Should all links for the same content be displayed here? or only one for file/folder? --> all links should be displayed in shared links view.
- [X] For all the things: ** grant compatibility with server <= 9.x **
- [X] Confirm remove link
- [x] Keyboards UI 
- [x] Clean unneeded code of the previous sharing version
- [x] [QA] Design test plan @jesmrec
- [ ] [QA] Validate test plan [WIP] @jesmrec

_____

BUGS & IMPROVEMENTS

- [X] (1) Shares not displayed https://github.com/owncloud/ios/issues/866#issuecomment-301397125 [FIXED] @nasli
- [X] (2) Allow uploads unreachable https://github.com/owncloud/ios/issues/866#issuecomment-301470426 [FIXED] @nasli
- [X] (3) Allow uploads can not be edited https://github.com/owncloud/ios/issues/866#issuecomment-301471382 [FIXED] @nasli
- [X] (4) Special characters in link name https://github.com/owncloud/ios/issues/866#issuecomment-301756902 [FIXED]  @nasli
- [X] (5) Password not reset https://github.com/owncloud/ios/issues/866#issuecomment-301795055 [FIXED] @nasli
- [X] (6) Date picker with selected date https://github.com/owncloud/ios/issues/866#issuecomment-301802983 [FIXED] @nasli
- [X] (7) Padding between expiration field and warning message https://github.com/owncloud/ios/issues/866#issuecomment-301814553 [WON'T FIX, future improvement maybe] @nasli
- [X] (8) Edit public link with password enforced https://github.com/owncloud/ios/issues/866#issuecomment-301816933 [FIXED] @nasli
- [X] (9) Warning is swipped and make the app crashes https://github.com/owncloud/ios/issues/866#issuecomment-301818149 [FIXED] @nasli
- [X] (10) Shared links display https://github.com/owncloud/ios/issues/866#issuecomment-301823697 [WON'T FIX, out of the scope the share link view tab]  @nasli
- [X] (11) UX in password enabling https://github.com/owncloud/ios/issues/866#issuecomment-302012004 [WON'T FIX, here]
- [X] (12) Become first response link name [WON'T FIX] @nasli  https://github.com/owncloud/ios/issues/866#issuecomment-302033705
- [X] (13) Expiration date and view refresh  https://github.com/owncloud/ios/issues/866#issuecomment-302051412 [FIXED] 
- [X] (14) Allow public uploads capability https://github.com/owncloud/ios/issues/866#issuecomment-302055894 [WONT'T FIX, core issue reported https://github.com/owncloud/core/issues/27922 ]
- [X] (15) Removal message in oC9 https://github.com/owncloud/ios/issues/866#issuecomment-302056162 [WON'T FIX, it is used same approach that in the server]
- [X] (16) Connection error message appears twice https://github.com/owncloud/ios/issues/866#issuecomment-302058975 [WONT'T FIX, here]
- [X] (17) Delete in oC 9 https://github.com/owncloud/ios/issues/866#issuecomment-302059955 [FIXED] 

Related on Android https://github.com/owncloud/android/issues/1928

PR: https://github.com/owncloud/ios/pull/887
library PR: https://github.com/owncloud/ios-library/pull/155

Fixes: #861

### PR DESCRIPTION
**Pull request by nasli**
Closes #866 
Fixes: #861

PR library https://github.com/owncloud/ios-library/pull/155

### CODE REVIEW

File: Owncloud iOs Client/Tabs/FileTab/Share/ShareFileOrFolder.m [hunk 1]
```diff
@@ -558,143 +121,42 @@ - (void) updateShareLink:(OCSharedDto *)ocShare withPassword:(NSString*)password
...
-            }
         }
-        if (!isSamlCredentialsError) {
-            
-            [self updateLocalShareLink:ocShare];
+        
+        if (isSamlCredentialsError) {
+            [self errorLogin];
+        } else {
+            [self.delegate sharelinkOptionsUpdated]; //TODO:return ocsharedDto instead of responsese data and not call sharelinkoptionsupdated
```
Comment (davivel) [inline Owncloud iOs Client/Tabs/FileTab/Share/ShareFileOrFolder.m]: Good feelings in that TODO 💯

File: Owncloud iOs Client/Tabs/FileTab/Share/ShareMainViewController.m [hunk 2]
```diff
@@ -332,237 +134,48 @@ - (void) setBarButtonStyle {
...
-
-
-- (void) passwordProtectedSwithValueChanged:(UISwitch*) sender{
-    
-     if (self.isPasswordProtectEnabled){
-
-        if (APP_DELEGATE.activeUser.hasCapabilitiesSupport == serverFunctionalitySupported) {
-            OCCapabilities *cap = APP_DELEGATE.activeUser.capabilitiesDto;
             
-            if (cap.isFilesSharingPasswordEnforcedEnabled) {
```
Comment (davivel) [inline Owncloud iOs Client/Tabs/FileTab/Share/ShareMainViewController.m]: We should have both classes ```OCShareUser```  and ```OCShareLink```, or none of them (and just use ```OCSharedDto```) ; but not only ```OCShareUser```.

But, definitely, not for now.

File: Owncloud iOs Client/Utils/ShareUtils.m [hunk 3]
```diff
@@ -32,7 +40,232 @@ + (NSMutableArray *) manageTheDuplicatedUsers: (NSMutableArray*) items{
...
+        return YES;
+    }
+    
+    return NO;
+}
+
+
+#pragma mark - Default value for link name
+
++ (NSString *) getDefaultLinkNameNormalicedOfFile:(FileDto *)file withLinkShares:(NSArray *)publicLinkShared {
```
Comment (davivel) [inline Owncloud iOs Client/Utils/ShareUtils.m]: Typo: ```Normalized```

File: Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+            [self updateEnabledOptions];
+        }
+//        _manageNetworkErrors = [ManageNetworkErrors new];
+//        _manageNetworkErrors.delegate = self;
+    }
+    
+    return self;
+}
+
+//- (void) dismissKeyboard {
```
Comment (davivel) [inline Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m]: Please, remove commented code if unneeded.

File: Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m [hunk 5]
```diff
@@ -0,0 +1,807 @@
...
+//    ShareLinkOptionCell *cellName = [self.shareLinkOptionsTableView viewWithTag:1];
+//    ShareLinkOptionCell *cellPassword = [self.shareLinkOptionsTableView viewWithTag:2];
+//    
+//    [cellName.optionTextField resignFirstResponder];
+//    [cellPassword.optionTextField resignFirstResponder];
+//}
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    // Do any additional setup after loading the view from its nib.
```
Comment (davivel) [inline Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m]: Same here.

File: Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+        NSArray *topLevelObjects = [[NSBundle mainBundle] loadNibNamed:shareLinkOptionNib owner:self options:nil];
+        shareLinkOptionCell = (ShareLinkOptionCell *)[topLevelObjects objectAtIndex:0];
+    }
+
+    [shareLinkOptionCell.optionSwith removeTarget:nil action:NULL forControlEvents:UIControlEventValueChanged];
+    shareLinkOptionCell.tag = indexPath.section+1;
+    
+    switch (indexPath.section) {
+            
+        case 0:
```
Comment (davivel) [inline Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m]: So much fun with TableViews... :(

File: Owncloud iOs Client/DataBase/Queries/ManageSharesDB.m [hunk 7]
```diff
@@ -164,6 +163,8 @@ + (NSMutableArray*) getSharesByFolder:(FileDto *) folder {
             sharedDto.shareWithDisplayName = [rs stringForColumn:@"share_with_display_name"];
             sharedDto.isDirectory = [rs boolForColumn:@"is_directory"];
             sharedDto.idRemoteShared = [rs intForColumn:@"id_remote_shared"];
```
Comment (davivel) [inline Owncloud iOs Client/DataBase/Queries/ManageSharesDB.m]: Let's do it, it's an easy refactoring.

File: Owncloud iOs Client/DataBase/Queries/ManageCapabilitiesDB.m [hunk 8]
```diff
@@ -55,7 +55,7 @@ +(OCCapabilities *) getCapabilitiesOfUserId:(NSInteger) userId{
     FMDatabaseQueue *queue = Managers.sharedDatabase;
     
     [queue inDatabase:^(FMDatabase *db) {
-        FMResultSet *rs = [db executeQuery:@"SELECT id, id_user, version_major, version_minor, version_micro, version_string, version_edition, core_poll_intervall, is_files_sharing_api_enabled, is_files_sharing_share_link_enabled, is_files_sharing_password_enforced_enabled, is_files_sharing_expire_date_by_default_enabled, is_files_sharing_expire_date_enforce_enabled, files_sharing_expire_date_days_number, is_files_sharing_allow_user_send_mail_notification_about_share_link_enabled, is_files_sharing_allow_public_uploads_enabled, is_files_sharing_allow_user_send_mail_notification_about_other_users_enabled, is_files_sharing_re_sharing_enabled, is_files_sharing_allow_user_send_shares_to_other_servers_enabled, is_files_sharing_allow_user_receive_shares_to_other_servers_enabled, is_file_big_file_chunking_enabled, is_file_undelete_enabled, is_file_versioning_enabled FROM capabilities WHERE id_user = ?", [NSNumber numberWithInteger:userId]];
+        FMResultSet *rs = [db executeQuery:@"SELECT * FROM capabilities WHERE id_user = ?", [NSNumber numberWithInteger:userId]];
```
Comment (nasli) [inline Owncloud iOs Client/DataBase/Queries/ManageCapabilitiesDB.m]: 👍

File: Owncloud iOs Client/DataBase/Queries/ManageSharesDB.m (hunk 7 above)
Comment (nasli) [inline Owncloud iOs Client/DataBase/Queries/ManageSharesDB.m]: 👍

Comment (nasli) [inline Owncloud iOs Client/DataBase/Queries/ManageSharesDB.m]: I rather leave it for next release too, qa already approved that, let's made the release..

File: Owncloud iOs Client/Network/Shares/CheckShares.m [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+//  Owncloud iOs Client
+//
+//  Created by Noelia Alvarez on 08/05/17.
+//
+//
+
+#import "CheckShares.h"
+#import "AppDelegate.h"
+
+@implementation CheckShares
```
Comment (nasli) [inline Owncloud iOs Client/Network/Shares/CheckShares.m]: For separate share request than were actually are. For now I can remove it

File: Owncloud iOs Client/Tabs/FileTab/Share/ShareMainViewController.m (hunk 2 above)
Comment (nasli) [inline Owncloud iOs Client/Tabs/FileTab/Share/ShareMainViewController.m]: Just OCSharedDto please..

File: Owncloud iOs Client/Utils/ShareUtils.m (hunk 3 above)
Comment (nasli) [inline Owncloud iOs Client/Utils/ShareUtils.m]: 👍

File: Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m (hunk 4 above)
Comment (nasli) [inline Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m]: 👍

File: Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m (hunk 6 above)
Comment (nasli) [inline Owncloud iOs Client/Tabs/FileTab/Share/ShareLinkViewController.m]: :sweat_smile:

File: Owncloud iOs Client/DataBase/Queries/ManageSharesDB.m (hunk 7 above)
Comment (davivel) [inline Owncloud iOs Client/DataBase/Queries/ManageSharesDB.m]: :+1:

**Comment by davivel** [comment]
:+1: now, just pending of Travis

**Pull request merged**